File: cli/src/commands/destroy.ts
import * as fs from 'fs';
import * as path from 'path';
import { dockerComposeOrThrow } from '@/docker.js';
import chalk from 'chalk';
import { loadConfig, configPath } from '@/config.js';
import { confirm } from '@/prompts.js';
import { removeEntry } from '@/image-registry.js';
import { resolveWorkspace, projectPaths } from '@/project.js';

import { parseCommonFlags, helpBlock } from '@/parse.js';
import type { Command } from '@/commands/command.js';

export interface DestroyFlags {
  yes: boolean;
  help: boolean;
  interactive: boolean;
}

export function parseDestroyFlags(argv: string[]): DestroyFlags {
  const { flags, remaining } = parseCommonFlags(argv);
  if (remaining.length > 0) {
    throw new Error(`Unknown flag for destroy: ${remaining[0]}`);
  }
  return flags;
}

export function destroyHelp(): string {
  return helpBlock(
    'devcontainer-cli destroy',
    'tear down everything for the current project',
    'devcontainer-cli destroy [flags]\n\nRuns \'docker compose down -v\' (containers + volumes), then deletes the generated\n.dc_<workspace>/ directory and devcontainer.config.json. This is irreversible.',
    [
      { name: '-y, --yes', description: 'Skip confirmation prompt' },
      { name: '--no-interactive', description: 'Non-interactive mode' },
      { name: '-h, --help', description: 'Show this help' },
    ]
  );
}

export async function runDestroy(argv: string[]): Promise<void> {
  const flags = parseDestroyFlags(argv);
  if (flags.help) { console.log(destroyHelp()); return; }

  const cwd = process.cwd();
  const workspace = resolveWorkspace(cwd);
  const paths = projectPaths(cwd, workspace);
  const projectDir = paths.projectDir;
  const composeFile = paths.composeFile;
  const cfgPath = configPath(cwd);

  if (!flags.yes) {
    if (!flags.interactive) {
      throw new Error("destroy is irreversible; pass --yes to confirm in non-interactive mode.");
    }
    const proceed = await confirm(
      'destroy',
      `Destroy '${workspace}'? Removes containers, volumes, .dc_${workspace}/ and devcontainer.config.json.`,
      false,
    );
    if (!proceed) { console.log(chalk.yellow('Cancelled.')); return; }
  }

  if (fs.existsSync(composeFile)) {
    console.log(chalk.yellow(`\nBringing down '${workspace}' (with volumes)...\n`));
    dockerComposeOrThrow(composeFile, ['down', '-v']);
  } else {
    console.log(chalk.gray(`No compose file at ${composeFile}; skipping 'docker compose down'.`));
  }

  if (fs.existsSync(projectDir)) {
    fs.rmSync(projectDir, { recursive: true, force: true });
    console.log(chalk.gray(`Removed ${projectDir}`));
  }
  if (fs.existsSync(cfgPath)) {
    fs.rmSync(cfgPath, { force: true });
    console.log(chalk.gray(`Removed ${cfgPath}`));
  }
  removeEntry(cwd);

  console.log(chalk.green.bold('\nDestroyed.\n'));
}

export const destroyCommand: Command<DestroyFlags> = {
  name: 'destroy',
  summary: 'down -v + delete .dc_<workspace>/ and config (irreversible)',
  parse: parseDestroyFlags,
  help: destroyHelp,
  run: runDestroy,
};

File: cli/internal/commands/destroy.go
package commands

import (
	"fmt"
	"os"

	"github.com/fatih/color"
	"github.com/joacohbc/my-devcontainer-installer/cli/internal/domain"
	"github.com/joacohbc/my-devcontainer-installer/cli/internal/infra/docker"
	"github.com/joacohbc/my-devcontainer-installer/cli/internal/infra/project"
	"github.com/joacohbc/my-devcontainer-installer/cli/internal/infra/prompt"
	"github.com/spf13/cobra"
)

func init() { register(newDestroyCommand()) }

func newDestroyCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "destroy",
		Short: "down -v + delete .dc_<workspace>/ and config (irreversible)",
		Long: `devcontainer-cli destroy — tear down everything for the current project

Runs 'docker compose down -v' (containers + volumes), then deletes the generated
.dc_<workspace>/ directory and devcontainer.config.json. This is irreversible.`,
		SilenceUsage: true,
		RunE:         runDestroy,
	}
	addYesFlag(cmd)
	addInteractiveFlag(cmd)
	return cmd
}

func runDestroy(cmd *cobra.Command, _ []string) error {
	cwd, _ := os.Getwd()
	cfg, _ := domain.LoadConfig(cwd)
	workspace := project.ResolveWorkspace(cwd, cfg)
	paths := project.ProjectPaths(cwd, workspace)
	cfgPath := domain.ConfigPath(cwd)

	if !yesFlag(cmd) {
		if !interactiveFlag(cmd) {
			return fmt.Errorf("destroy is irreversible; pass --yes to confirm in non-interactive mode")
		}
		proceed, err := prompt.Confirm(
			fmt.Sprintf("Destroy '%s'? Removes containers, volumes, .dc_%s/ and devcontainer.config.json.", workspace, workspace),
			false,
		)
		if err != nil {
			return err
		}
		if !proceed {
			color.Yellow("Cancelled.")
			return nil
		}
	}

	if _, err := os.Stat(paths.ComposeFile); err == nil {
		color.Yellow("\nBringing down '%s' (with volumes)...\n", workspace)
		if err := docker.DockerComposeOrThrow(paths.ComposeFile, []string{"down", "-v"}, nil); err != nil {
			return err
		}
	} else {
		color.New(color.FgWhite).Printf("No compose file at %s; skipping 'docker compose down'.\n", paths.ComposeFile)
	}

	if _, err := os.Stat(paths.ProjectDir); err == nil {
		if err := os.RemoveAll(paths.ProjectDir); err == nil {
			color.New(color.FgWhite).Printf("Removed %s\n", paths.ProjectDir)
		}
	}
	if _, err := os.Stat(cfgPath); err == nil {
		if err := os.Remove(cfgPath); err == nil {
			color.New(color.FgWhite).Printf("Removed %s\n", cfgPath)
		}
	}
	domain.RemoveEntry(cwd)

	color.New(color.FgGreen, color.Bold).Print("\nDestroyed.\n\n")
	return nil
}
